### PR TITLE
Producer support schema updates

### PR DESCRIPTION
**Pull request by OrMemphis**

### CODE REVIEW

**Comment by OrMemphis** [comment]
Tests fail because the master broker doesn't yet send the new producerCreation responses

File: producer.go [hunk 1]
```diff
@@ -115,6 +186,9 @@ func (p *Producer) getDestructionReq() any {
 
 // Destroy - destoy this producer.
 func (p *Producer) Destroy() error {
+	if err := p.conn.removeSchemaUpdatesListener(p.stationName); err != nil {
+		panic(err)
```
Comment (idanasulinmemphis) [inline producer.go:190]: panic is not good I think since it will cause the user application to crash

**Review by idanasulinmemphis — DISMISSED** [review]

File: producer.go [hunk 2]
```diff
@@ -255,10 +256,15 @@ func (opts *ProduceOpts) produce(p *Producer) error {
 	opts.MsgHeaders.MsgHeaders["$memphis_connectionId"] = []string{p.conn.ConnId}
 	opts.MsgHeaders.MsgHeaders["$memphis_producedBy"] = []string{p.Name}
 
+	data, err := p.validateMsg(opts.Message)
```
Comment (idanasulinmemphis) [inline producer.go:259]: check wether you need to validate, otherwise a lot of functions will be called

File: producer.go [hunk 3]
```diff
@@ -279,6 +285,38 @@ func (opts *ProduceOpts) produce(p *Producer) error {
...
+			return nil, errors.New("Unsupported message type")
+		}
+
+	}
+
+	return sd.validateProtoMsg(msg)
+}
+
+func (p *Producer) getSchemaDetails() (schemaDetails, error) {
+	return p.conn.getSchemaDetails(p.stationName)
```
Comment (idanasulinmemphis) [inline producer.go:308]: please make sure that your map keys are always internally names here and everywhere else

File: producer.go [hunk 4]
```diff
@@ -279,6 +285,38 @@ func (opts *ProduceOpts) produce(p *Producer) error {
 	}
 }
 
+func (p *Producer) validateMsg(msg any) ([]byte, error) {
+	sd, err := p.getSchemaDetails()
+	if err != nil {
+		return nil, err
+	}
+
+	if sd.schemaType == "" {
```
Comment (idanasulinmemphis) [inline producer.go:294]: in case no schema found:
add comment this is the case and validation is not needed

File: producer.go [hunk 5]
```diff
@@ -279,6 +285,38 @@ func (opts *ProduceOpts) produce(p *Producer) error {
...
+		switch msg.(type) {
+		case []byte:
+			return msg.([]byte), nil
+		default:
+			return nil, errors.New("Unsupported message type")
+		}
+
+	}
+
+	return sd.validateProtoMsg(msg)
```
Comment (idanasulinmemphis) [inline producer.go:304]: You should support other types as well so please make it in switch case base on the schema typ

File: producer.go [hunk 6]
```diff
@@ -279,6 +285,38 @@ func (opts *ProduceOpts) produce(p *Producer) error {
...
+}
+
+func (p *Producer) getSchemaDetails() (schemaDetails, error) {
+	return p.conn.getSchemaDetails(p.stationName)
+}
+
+func (p *Producer) PrintSchemaDetails() {
```
Comment (idanasulinmemphis) [inline producer.go:311]: why print function?

File: station.go [hunk 7]
```diff
@@ -336,17 +343,84 @@ func (sus *stationUpdateSub) schemaUpdatesHandler(lock *sync.RWMutex) {
...
+		return nil, errors.New("Unsupported message type")
+	}
+
+	protoMsg := dynamicpb.NewMessage(sd.msgDescriptor)
+	err = proto.Unmarshal(msgBytes, protoMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	b := protoMsg.GetUnknown()
```
Comment (idanasulinmemphis) [inline station.go:420]: you don't need this one

File: producer.go (hunk 2 above)
Comment (OrMemphis) [reply producer.go:259]: there is a a check quite early inside that funciton
```
func (p *Producer) validateMsg(msg any) ([]byte, error) {
	sd, err := p.getSchemaDetails()
	if err != nil {
		return nil, err
	}

	if sd.schemaType == "" {
		switch msg.(type) {
		case []byte:
			return msg.([]byte), nil
		default:
			return nil, errors.New("Unsupported message type")
		}

	}

	return sd.validateProtoMsg(msg)
}

```

**Review by idanasulinmemphis — APPROVED** [review]

**Review by idanasulinmemphis — APPROVED** [review]

**Pull request merged**